lib/ecto/query/planner.ex

```diff
@@ -1913,8 +1914,13 @@ defmodule Ecto.Query.Planner do
   end
 
   defp type!(kind, query, expr, schema, field, allow_virtuals?) when is_atom(schema) do
+    only_writable? = kind in @write_only_kinds
```

Comment:
```suggestion
    type_key = if kind in @write_only_kinds, do: :writable_type, else: :type
```

And then pass it as first argument to the type call?